sys/arm64/arm64/copyinout.S

```diff
@@ -224,3 +224,128 @@ ending:
 	mov	x0, xzr		/* return 0 */
 	ret
 	.size	copycommon, . - copycommon
+
+#if __has_feature(capabilities)
+/*
+ * Copy specified amount of data from the user to kernel space, preserving
+ * capability tags
+ *
+ * int copyincap(const void *uaddr, void *kdaddr, size_t len)
+ */
+ENTRY(copyincap)
+	cbz	x2, 1f
+	adds	x3, x0, x2
+	b.cs 	copyio_fault_nopcb
+	ldr	x4, =VM_MAXUSER_ADDRESS
+	cmp	x3, x4
+	b.hi	copyio_fault_nopcb
+
+	/* Check the length is greater than 1 capability */
+	cmp	x2, #16
+	b.lo	copycommon
+	/* And the addresses will both align on a capability */
+	eor	x3, x0, x1		/* Compare the pointers */
+	and	x3, x3, #0xf		/* Check the lower bits */
+	cbnz	x3, copycommon		/* If different don't copy tags */
+
+	b	copycapcommon
+
+1:	mov	x0, xzr		/* return 0 */
+	ret
+END(copyincap)
+
+/*
+ * Copy specified amount of data from kernel to the user space, preserving
+ * capability tags
+ *
+ * int copyoutcap(const void *kaddr, void *udaddr, size_t len)
+ */
+ENTRY(copyoutcap)
+	cbz	x2, 1f
+	adds	x3, x1, x2
+	b.cs 	copyio_fault_nopcb
+	ldr	x4, =VM_MAXUSER_ADDRESS
+	cmp	x3, x4
+	b.hi	copyio_fault_nopcb
+
+	/* Check the length is greater than 1 capability */
+	cmp	x2, #16
+	b.lo	copycommon
+	/* And the addresses will both align on a capability */
+	eor	x3, x0, x1		/* Compare the pointers */
+	and	x3, x3, #0xf		/* Check the lower bits */
+	cbnz	x3, copycommon		/* If different don't copy tags */
+
+	b	copycapcommon
+
+1:	mov	x0, xzr		/* return 0 */
+	ret
+END(copyoutcap)
+
+/*
+ * Local helper
+ *
+ * x0 - src pointer
+ * x1 - dst pointer
```

Comment:
Both need to be capabilities in order for both copyincap and copyoutcap to have the right semantics.

Reply:
I guess using the integer-address loads/stores is fine for now as long as there is a follow-up change to make them use capabilities.